src/util/typed.rs

```diff
@@ -0,0 +1,2344 @@
+        Some(x) => x.as_str(),
+        None => return results,
+    };
+
+    if results.iter().any(|x| x == primary_type) || !types.contains_key(primary_type) {
+        return results;
+    }
+
+    results.push(primary_type.into());
+
+    for field in &types[primary_type] {
+        for dep in find_type_dependencies(&field.kind, types, results.clone()) {
+            if results.iter().all(|x| x != &dep) {
+                results.push(dep);
+            }
+        }
+    }
+
+    results
+}
+
+type TypeName = String;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct SanitizedType {
+    name: String,
+    #[serde(rename = "type")]
+    kind: String,
+}
+
+#[derive(Debug, Clone)]
+struct SanitizedData {
+    types: HashMap<TypeName, Vec<SanitizedType>>,
+    domain: Value,
+    primary_type: String,
+    message: Value,
+}
+
+fn sanitize(typed_data: Value) -> Result<SanitizedData, failure::Error> {
+    log::trace!("sanitizing data: {:#?}", typed_data);
+    let data = typed_data
+        .as_object()
+        .ok_or_else(|| failure::err_msg("TypedData parameter must be a JSON Object"))?;
+
+    let primary_type = data
+        .get("primaryType")
+        .ok_or_else(|| failure::err_msg("primaryType must not present in the TypedData object"))?;
+    let primary_type: String = serde_json::from_value(primary_type.clone())
+        .map_err(|err| failure::format_err!("failed to deserialize primaryType: {}", err))?;
+
+    let domain = data
+        .get("domain")
+        .ok_or_else(|| failure::err_msg("domain must be present in the TypedData object"))?;
+    let domain = domain.clone();
+
+    let message = data
+        .get("message")
+        .ok_or_else(|| failure::err_msg("message must be present in the TypedData object"))?;
+    let message = message.clone();
+
+    let types = data
+        .get("types")
+        .ok_or_else(|| failure::err_msg("types must be present in the TypedData object"))?;
+    let types: HashMap<TypeName, Vec<SanitizedType>> = serde_json::from_value(types.clone())
+        .map_err(|err| failure::format_err!("failed to deserialize types: {}", err))?;
+
+    let data = SanitizedData {
+        types,
+        domain,
+        primary_type,
+        message,
+    };
+
+    log::trace!("sanitized data: {:#?}", data);
+
+    Ok(data)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::util::typed::{encode_type, find_type_dependencies, SanitizedData};
+    use bitcoin::util::misc::hex_bytes;
+    use std::collections::HashMap;
+
+    mod hash {
+        use crate::util::typed::hash;
+
+        fn do_test_hash(typed_data: &str, expected: &str) {
+            let input =
+                serde_json::from_str(typed_data).expect("failed to deserialize 'typed_data' json");
+            let got = hash(input).unwrap();
+            let got_base64 = base64::encode(&got);
+            assert_eq!(got_base64, expected);
+        }
+
+        #[test]
+        fn test_1() {
```

Comment:
Better to use meaningfull test name - description what these test checks.

Reply:
Tests are literally a dump of input/output from functions in eth-sig js library